File: utils/signature.js
// Base64编码类
class Base64 {
    constructor() {
        this._keyStr = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
    }
    
    encode(input) {
        let output = "";
        let chr1, chr2, chr3, enc1, enc2, enc3, enc4;
        let i = 0;
        input = this._utf8_encode(input);
        while (i < input.length) {
            chr1 = input.charCodeAt(i++);
            chr2 = input.charCodeAt(i++);
            chr3 = input.charCodeAt(i++);
            enc1 = chr1 >> 2;
            enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
            enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
            enc4 = chr3 & 63;
            if (isNaN(chr2)) {
                enc3 = enc4 = 64;
            } else if (isNaN(chr3)) {
                enc4 = 64;
            }
            output = output +
                this._keyStr.charAt(enc1) + this._keyStr.charAt(enc2) +
                this._keyStr.charAt(enc3) + this._keyStr.charAt(enc4);
        }
        return output;
    }
    
    _utf8_encode(string) {
        string = string.replace(/\r\n/g,"\n");
        let utftext = "";
        for (let n = 0; n < string.length; n++) {
            let c = string.charCodeAt(n);
            if (c < 128) {
                utftext += String.fromCharCode(c);
            } else if((c > 127) && (c < 2048)) {
                utftext += String.fromCharCode((c >> 6) | 192);
                utftext += String.fromCharCode((c & 63) | 128);
            } else {
                utftext += String.fromCharCode((c >> 12) | 224);
                utftext += String.fromCharCode(((c >> 6) & 63) | 128);
                utftext += String.fromCharCode((c & 63) | 128);
            }
        }
        return utftext;
    }
}

// 自定义签名函数
function customSignature(str, secret) {
    let s = str + secret;
    // 步长抽样（每隔5个字符取1个）
    let sampled = "";
    for (let i = 0; i < s.length; i += 5) {
        sampled += s[i];
    }
    // 反转+charCode偏移
    let arr = [];
    for (let i = sampled.length - 1; i >= 0; i--) {
        arr.push(String.fromCharCode(sampled.charCodeAt(i) + 2));
    }
    let result = arr.join("");
    // Base64 编码
    let b64 = new Base64().encode(result);
    // 截取前32字符
    return b64.slice(0, 32);
}

// 生成带签名的URL
function generateSignedUrl(baseUrl, params, secret = "default_secret") {
    // 构建query字符串（排除signature参数）
    const queryParams = new URLSearchParams();
    for (const [key, value] of Object.entries(params)) {
        if (key !== 'signature') {
            queryParams.append(key, value);
        }
    }
    
    const queryString = queryParams.toString();
    const signature = customSignature(queryString, secret);
    
    // 添加签名参数
    queryParams.append('signature', signature);
    
    return baseUrl + '?' + queryParams.toString();
}

// 验证签名
function verifySignature(queryString, signature, secret = "default_secret") {
    const expectedSignature = customSignature(queryString, secret);
    return signature === expectedSignature;
}

// 从URL中提取query字符串
function getQueryStringFromUrl(url) {
    const urlObj = new URL(url);
    return urlObj.search.substring(1); // 去掉开头的'?'
}

// 导出函数（如果在Node.js环境中使用）
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        Base64,
        customSignature,
        generateSignedUrl,
        verifySignature,
        getQueryStringFromUrl
    };
}

File: utils/signature.py
import base64
import os
from typing import Optional


def custom_signature(query_string: str, secret: str) -> str:
    """
    生成自定义签名
    :param query_string: URL的query字符串
    :param secret: 密钥
    :return: 签名结果
    """
    # 拼接query和secret
    s = query_string + secret
    
    # 步长抽样（每隔5个字符取1个）
    sampled = ""
    for i in range(0, len(s), 5):
        sampled += s[i]
    
    # 反转+charCode偏移
    arr = []
    for i in range(len(sampled) - 1, -1, -1):
        arr.append(chr(ord(sampled[i]) + 2))
    
    result = "".join(arr)
    
    # Base64编码
    b64 = base64.b64encode(result.encode('utf-8')).decode('utf-8')
    
    # 截取前32字符
    return b64[:32]


def verify_signature(query_string: str, signature: str, secret: str = None) -> bool:
    """
    验证签名
    :param query_string: URL的query字符串
    :param signature: 要验证的签名
    :param secret: 密钥，如果为None则从环境变量获取
    :return: 验证结果
    """
    if secret is None:
        secret = os.getenv("PARSE_VIDEO_SECRET", "default_secret")
    
    expected_signature = custom_signature(query_string, secret)
    return signature == expected_signature


def get_query_string_from_url(url: str) -> str:
    """
    从URL中提取query字符串
    :param url: 完整的URL
    :return: query字符串
    """
    from urllib.parse import urlparse
    parsed_url = urlparse(url)
    return parsed_url.query


def generate_signed_url(base_url: str, params: dict, secret: str = None) -> str:
    """
    生成带签名的URL
    :param base_url: 基础URL
    :param params: 参数字典
    :param secret: 密钥
    :return: 带签名的完整URL
    """
    if secret is None:
        secret = os.getenv("PARSE_VIDEO_SECRET", "default_secret")
    
    # 构建query字符串（排除signature参数）
    query_params = []
    for key, value in sorted(params.items()):
        if key != 'signature':
            query_params.append(f"{key}={value}")
    
    query_string = "&".join(query_params)
    signature = custom_signature(query_string, secret)
    
    # 添加签名参数
    query_params.append(f"signature={signature}")
    
    return f"{base_url}?{'&'.join(query_params)}"
